File: dsa/linkedlist/DoublyLinkedList.java
package linkedlist;

import java.util.Scanner;

public class DoublyLinkedList {
    static Scanner sc = new Scanner(System.in);
    static Node head;
    static Node tail;

    static class Node {
        int data;
        Node next;
        Node prev;

        Node(int data) {
            this.data = data;
            this.next = null;
            this.prev = null;
        }
    }

    public static void main(String[] args) {
        while (true) {
            System.out.print("Press:\n" +
                    "1 to Insert\n" +
                    "2 to Display\n" +
                    "3 to Delete\n" +
                    "4 to Delete All\n" +
                    "5 to exit()\n" +
                    "Enter your Choice -> ");
            int Choice = sc.nextInt();
            switch (Choice) {
                case 1:
                    insert();
                    break;
                case 2:
                    display();
                    break;
                case 3:
                    delete();
                    break;
                case 4:
                    deleteAll();
                    break;
                case 5:
                    System.exit(0);
                default:
                    System.out.println("Wrong Choice! ");
            }
        }
    }

    static void insert() {
        System.out.print("Enter your Data -> ");
        int data = sc.nextInt();
        Node newNode = new Node(data);
        if (head == null) {
            head = newNode;
            tail = newNode;
            System.out.println("Data Inserted!");
            return;
        }
        Node currentNode = head;
        while (currentNode.next != null)
            currentNode = currentNode.next;
        currentNode.next = newNode;
        newNode.prev = currentNode;
        tail = newNode;
        System.out.println("Data Inserted!");
    }

    static void delete() {
        if (head == null) {
            System.out.println("List is Empty!");
            return;
        }
        System.out.println("Deleted -> " + head.data);
        if (head.next == null)
            head = null;
        else {
            head = head.next;
            head.prev = null;
        }
    }

    static void display() {
        if (head == null) {
            System.out.println("List is Empty!");
            return;
        }
        // forward
        Node currentNode;
        for (currentNode = head; currentNode != null; currentNode = currentNode.next) {
            System.out.print(currentNode.data + " ");
        }
        System.out.println();
        // reverse
        for (currentNode = tail; currentNode != null; currentNode = currentNode.prev)
            System.out.print(currentNode.data + " ");
        System.out.println();
    }

    static void deleteAll() {
        if (head == null) {
            System.out.println("List is Empty!");
            return;
        }
        System.out.println("List Successfully Cleared!");
        head = null;
    }
}

File: Python/linked_list/doubly_linked_list.py
from node import Node


class DoublyLinkedList:
    def __init__(self):
        self.head = self.tail = None
        self.size = 0

    def is_empty(self) -> bool:
        return self.head is None and self.tail is None

    def insert_begin(self, data):
        new_node = Node(data)
        if self.is_empty():
            self.head = self.tail = new_node
        else:
            new_node.next = self.head
            self.head.prev = new_node
            self.head = new_node
        print("Data Inserted!")
        self.size += 1

    def insert_at_pos(self, data, pos: int):
        if pos == 0:
            print("Invalid Input! Index Starts from 1.")
        elif pos == 1:
            self.insert_begin(data)
        elif pos > self.size:
            print(f"Invalid Input! List size is: {self.size}")
        else:
            new_node = Node(data)
            current_node = self.head
            for _ in range(0, pos - 2):
                current_node = current_node.next
            new_node.next = current_node.next
            current_node.next.prev = new_node
            current_node.next = new_node
            new_node.prev = current_node
            print("Data Inserted!")
            self.size += 1

    def insert_at_last(self, data):
        new_node = Node(data)
        if self.is_empty():
            self.head = self.tail = new_node
        else:
            self.tail.next = new_node
            new_node.prev = self.tail
            self.tail = new_node
        print("Data Inserted!")
        self.size += 1

    def delete_begin(self):
        if self.is_empty():
            print("List is Empty!")
        elif self.head.next is None:
            self.head = self.tail = None
            print("List Cleared Successfully!")
        else:
            print(f"Successfully Deleted '{self.head.data}' from beginning.")
            self.head = self.head.next
            self.head.prev = None
            self.size -= 1

    def delete_at_pos(self, pos: int):
        if self.is_empty():
            print("List is Empty!")
        elif pos == 0:
            print("Invalid Input! Index Starts from 1.")
        elif pos == 1:
            self.delete_begin()
        elif pos > self.size:
            print(f"Invalid Input! List size is: {self.size}")
        else:
            current_node = self.head
            for _ in range(0, pos - 2):
                current_node = current_node.next
            print(
                f"Successfully Deleted '{current_node.next.data}' from position '{pos}'."
            )
            current_node.next = current_node.next.next
            current_node.next.prev = current_node
            self.size -= 1

    def delete_at_last(self):
        if self.is_empty():
            print("List is Empty!")
        elif self.head.next is None:
            self.head = self.tail = None
            print("List Cleared Successfully!")
            self.size = 0
        else:
            print(f"Successfully Deleted '{self.tail.data}' from last.")
            self.tail = self.tail.prev
            self.tail.next = None
            self.size -= 1

    def delete_all(self):
        if self.is_empty():
            print("List is Empty!")
        else:
            print("List Cleared Successfully!")
            self.head = self.tail = None
            self.size = 0

    def display(self):
        if self.is_empty():
            print("List is Empty!")
        else:
            current_node = self.head
            print("[", end="")
            while current_node is not None:
                print(current_node.data, end=", ")
                current_node = current_node.next
            print("\b\b]")
            print(f"List size is -> {self.size}")

    def display_reverse(self):
        if self.is_empty():
            print("List is Empty!")
        else:
            current_node = self.tail
            print("[", end="")
            while current_node is not None:
                print(current_node.data, end=", ")
                current_node = current_node.prev
            print("\b\b]")
            print(f"List size is -> {self.size}")
